subsys/net/ip/ipv4_autoconf.c

```diff
@@ -0,0 +1,288 @@
+/** @file
+ * @brief IPv4 autoconf related functions
+ */
+
+/*
+ * Copyright (c) 2016 Matthias Boesl
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if defined(CONFIG_NET_DEBUG_IF)
+#define SYS_LOG_DOMAIN "net/ipv4_autoconf"
+#define NET_DEBUG 1
+#endif
+
+#include "net_private.h"
+#include <errno.h>
+#include <net/arp.h>
+#include <net/nbuf.h>
+#include <net/net_core.h>
+#include <net/net_if.h>
+
+#include <net/ipv4_autoconf.h>
+
+/* initial random delay*/
+#define IPV4_AUTOCONF_PROBE_WAIT 1
+/* number of probe packets */
+#define IPV4_AUTOCONF_PROBE_NUM 3
+/* minimum delay till repeated probe */
+#define IPV4_AUTOCONF_PROBE_MIN 1
+/* maximum delay till repeated probe */
+#define IPV4_AUTOCONF_PROBE_MAX 2
+/* number of announcement packets */
+#define IPV4_AUTOCONF_ANNOUNCE_NUM 2
+/* time between announcement packets */
+#define IPV4_AUTOCONF_ANNOUNCE_INTERVAL 2
+/* max conflicts before rate limiting */
+#define IPV4_AUTOCONF_MAX_CONFLICTS 10
+/* delay between successive attempts */
+#define IPV4_AUTOCONF_RATE_LIMIT_INTERVAL 60
+/* minimum interval between defensive ARPs */
+#define IPV4_AUTOCONF_DEFEND_INTERVAL 10
+
+static void send_probe(struct net_if *iface);
+static void send_announcement(struct net_if *iface);
+
+static inline struct net_buf *prepare_arp(struct net_if *iface)
+{
+	struct net_buf *buf, *frag;
+	struct net_arp_hdr *hdr;
+	struct net_eth_hdr *eth;
+
+	buf = net_nbuf_get_reserve_tx(0, K_FOREVER);
+	if (!buf) {
+		goto fail;
+	}
+
+	frag = net_nbuf_get_reserve_data(sizeof(struct net_eth_hdr), K_FOREVER);
+	if (!frag) {
+		goto fail;
+	}
+
+	net_buf_frag_add(buf, frag);
+	net_nbuf_set_iface(buf, iface);
+	net_nbuf_set_family(buf, AF_INET);
+	net_nbuf_set_ll_reserve(buf, sizeof(struct net_eth_hdr));
+
+	eth = NET_ETH_BUF(buf);
+	hdr = NET_ARP_BUF(buf);
+	eth->type = htons(NET_ETH_PTYPE_ARP);
+
+	/* ethernet header */
+	memcpy(&eth->src.addr, net_if_get_link_addr(iface)->addr,
+	       sizeof(struct net_eth_addr));
+	memset(&eth->dst.addr, 0xFF, sizeof(struct net_eth_addr));
+
+	hdr->hwtype = htons(NET_ARP_HTYPE_ETH);
+	hdr->protocol = htons(NET_ETH_PTYPE_IP);
+	hdr->hwlen = sizeof(struct net_eth_addr);
+	hdr->protolen = sizeof(struct in_addr);
+	hdr->opcode = htons(NET_ARP_REQUEST);
+
```

Comment:
all the above is not necessary, this is handled in susbsys/net/ip/l2/: arp.c and ethernet.c